ntoskrnl/ex/time.c

```diff
@@ -30,6 +30,108 @@ ULONG ExpTimerResolutionCount = 0;
 
 /* FUNCTIONS ****************************************************************/
 
+/*++
+ * If successful, this function sets the following global variables:
+ * ExpTimeZoneInfo
+ * ExpTimeZoneBias
+ *--*/
+static
+BOOLEAN
+ExpGetTimeZoneId(
+    _In_ PLARGE_INTEGER TimeNow,
+    _Out_ PULONG TimeZoneId)
+{
+    LARGE_INTEGER StandardTime;
+    LARGE_INTEGER DaylightTime;
+    LARGE_INTEGER LocalTimeNow = *TimeNow;
+    NTSTATUS Status;
+    static BOOLEAN InitialRun = TRUE;
+
+    /* Read time zone information from the registry */
+    Status = RtlQueryTimeZoneInformation(&ExpTimeZoneInfo);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlQueryTimeZoneInformation() failed (Status 0x%08lx)\n", Status);
+        return FALSE;
+    }
+
+    /* Get the default bias */
+    ExpTimeZoneBias.QuadPart = (LONGLONG)ExpTimeZoneInfo.Bias * TICKSPERMINUTE;
+    if (!InitialRun)
+    {
+        /* Adjust TimeNow for Time Zone Bias to use UTC for comparisons */
+        LocalTimeNow.QuadPart -= ExpTimeZoneBias.QuadPart;
+    }
+    else
+    {
+        InitialRun = FALSE;
+    }
```

Comment:
I don't really like the use of a static local and different behavior based on when the function is called. Instead I suggest to modify the initial caller, so that it adds ExpTimeZoneBias to the time it passes to this function, so that the logic is right again.

Reply:
Removed static ```InitialRun``` variable and this whole section of code. Replaced it with calculating the Time Zone Bias before calling this function and accounting for it before making the call, so it comes in with the input parameters.